### ISSUE TITLE
`AuthUserPoolException` when using a custom tokenProvider

### ISSUE THREAD

**Issue by ryanwalters** [aws-amplify/amplify-js#13018]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

Next.js

### Amplify APIs

Authentication

### Amplify Version

v6

### Amplify Categories

auth

### Backend

None

### Environment information

<details>

`npx envinfo` did not work on my system. The below is using `npx --no-install next info`:

```
# Put output below this line
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Pro
Binaries:
  Node: 20.11.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.1.0
  eslint-config-next: 14.1.0
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
```

</details>

### Describe the bug

Our use case requires writing cookies for all subdomains for a given domain (e.g. `.domain.com`). Amplify v6 does not appear to allow us to configure this the same way v5 did via the `cookieStorage` option. `Amplify.configure()` seems to offer a potential workaround by means of configuring a custom `tokenProvider` and `credentialsProvider`.

Please note that at this point in time, I am only trying to get this working client side since operations like `signIn`, `signOut`, etc are not supported in server actions [as stated here](https://github.com/aws-amplify/amplify-js/issues/12827#issuecomment-1885273494).

However, after configuring these I receive the following error:
```
AuthUserPoolException: Auth UserPool not configured.
    at eval (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/core/dist/esm/errors/createAssertionFunction.mjs:14:19)
    at assertTokenProviderConfig (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/core/dist/esm/singleton/Auth/utils/index.mjs:23:69)
    at DefaultTokenStore.getAuthKeys (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/tokenProvider/TokenStore.mjs:129:101)
    at DefaultTokenStore.getDeviceMetadata (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/tokenProvider/TokenStore.mjs:109:37)
    at TokenOrchestrator.getDeviceMetadata (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/tokenProvider/TokenOrchestrator.mjs:132:37)
    at handlePasswordVerifierChallenge (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/utils/signInHelpers.mjs:500:52)
    at async retryOnResourceNotFoundException (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/utils/signInHelpers.mjs:796:16)
    at async signInWithSRP (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/apis/signInWithSRP.mjs:53:87)
    at async handleSignIn (webpack-internal:///(app-pages-browser)/./app/page.tsx:19:36)
```

### Expected behavior

User can sign in using a custom token provider.

### Reproduction steps

I've created a reproduction repo that showcases this issue. Please see reproduction steps outlined in the readme for "Issue 1" found in https://github.com/ryanwalters/amplify-6.

### Log output

<details>

```
// Put your logs below this line


```

</details>

**Comment by cwomack** [comment]
Hello, @ryanwalters and thanks for opening this issue.  Greatly appreciate you creating a sample repo as well! 

With the `AuthUserPoolException` error that you're experiencing, I believe that's being thrown because the code in your `useAuth.ts` file is going to be run server-side.  If you're Next.JS but not wrapping the code in that file in a client scope, it's being evaluated on the server side.

Also, I see the scoped config for Auth in the sample app that you provided.  Is that the same for your main app as well (not sure if they are the same repo/app)?  Or are you using the auto-generated config from the CLI?

**Comment by ryanwalters** [comment]
> With the AuthUserPoolException error that you're experiencing, I believe that's being thrown because the code in your useAuth.ts file is going to be run server-side. If you're Next.JS but not wrapping the code in that file in a client scope, it's being evaluated on the server side.

I've marked `useAuth.ts` with the `'use client'` directive. This change has been pushed to the sample repo. 

I now get a different error:

```
page.tsx:19 AuthUserPoolException: Auth UserPool not configured.
    at eval (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/core/dist/esm/errors/createAssertionFunction.mjs:14:19)
    at assertTokenProviderConfig (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/core/dist/esm/singleton/Auth/utils/index.mjs:23:69)
    at DefaultTokenStore.getAuthKeys (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/tokenProvider/TokenStore.mjs:129:101)
    at DefaultTokenStore.getDeviceMetadata (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/tokenProvider/TokenStore.mjs:109:37)
    at TokenOrchestrator.getDeviceMetadata (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/tokenProvider/TokenOrchestrator.mjs:132:37)
    at handlePasswordVerifierChallenge (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/utils/signInHelpers.mjs:500:52)
    at async retryOnResourceNotFoundException (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/utils/signInHelpers.mjs:796:16)
    at async signInWithSRP (webpack-internal:///(app-pages-browser)/./node_modules/@aws-amplify/auth/dist/esm/providers/cognito/apis/signInWithSRP.mjs:53:87)
    at async handleSignIn (webpack-internal:///(app-pages-browser)/./app/page.tsx:19:36)
```

Maybe this is progress? 😄 

> Also, I see the scoped config for Auth in the sample app that you provided. Is that the same for your main app as well (not sure if they are the same repo/app)? Or are you using the auto-generated config from the CLI?

Yes, we are using a pool that is created and configured in a separate application (as part of a larger CloudFormation stack rather than Amplify).

**Comment by JoeyOverby** [comment]
I'm having a similar issue where I get this error if I have a credentialsProvider in my configure call. 

This works fine (but can't get custom creds of course....) 
```
Amplify.configure(amplifyconfig);
``` 

But since I need some creds from Cognito and others from Auth0, when I do this it fails 
```
const customCredentialsProvider = AuthApi.getCustomCredsProvider();
Amplify.configure(amplifyconfig, {
  Auth: {
    credentialsProvider: customCredentialsProvider,
  },
});

```

With the second set of code it fails on the non redirected/normal sign in call. 
```
await signIn({
        username: email,
        password: "-",
      });
```

Error: `"Auth UserPool not configured."` 
Recovery Suggestion: `"Make sure to call Amplify.configure in your app with userPoolId and userPoolClientId."` 


Clearly I have those there as the configure call without the credentialsProvider works fine. But by adding the credentials provider it fails to load until after the user has signed in. (Which they can't - because sign in fails...)

**Comment by cwomack** [comment]
@ryanwalters, can you detail how you're defining `Amplify.configure()` call as well as how you've set up your custom tokenProvider?  Amplify.configure takes in 2 things, the resource config and then the library configs. As soon as you've done all the configuring, can you do the following to log out the config to see if anything unexpected is happening: 

 `console.log(Amplify.getConfig)`
 
 CC: @JoeyOverby, can you do the same logging as well?  Thanks!

**Comment by ryanwalters** [comment]
@cwomack The `Amplify.configure()` and token provider definitions can be found in the sample repo here:
https://github.com/ryanwalters/amplify-6/blob/main/hooks/useAuth.ts#L12-L49

Let me know if you'd like me to paste them here instead.

The output of `console.log(Amplify.getConfig())`:
```
{
  Auth: {
    Cognito: {
      userPoolId: 'us-east-1_***',
      userPoolClientId: '***'
    }
  }
}
```

**Comment by cwomack** [comment]
@ryanwalters After looking again at the code example in your sample repo, you are configuring the `tokenProvider` and `credentialsProvider` for the server side use cases. However, other parts of your code are only using Amplify on the **_client_** side. Can you clarify if you’re calling any API’s on the server side?

If you are only intending to use Amplify on the client side, [this document](https://docs.amplify.aws/gen1/react/build-a-backend/auth/manage-user-session/#update-your-token-saving-mechanism:~:text=Cookie%20Storage-,CookieStorage,-saves%20the%20tokens) will show you how to configure the `tokenProviders` to change the cookie domain.

Let us know if this helps clear things up!

**Comment by jeff2013** [comment]
> I'm having a similar issue where I get this error if I have a credentialsProvider in my configure call.
> 
> This works fine (but can't get custom creds of course....)
> 
> ```
> Amplify.configure(amplifyconfig);
> ```
> 
> But since I need some creds from Cognito and others from Auth0, when I do this it fails
> 
> ```
> const customCredentialsProvider = AuthApi.getCustomCredsProvider();
> Amplify.configure(amplifyconfig, {
>   Auth: {
>     credentialsProvider: customCredentialsProvider,
>   },
> });
> ```
> 
> With the second set of code it fails on the non redirected/normal sign in call.
> 
> ```
> await signIn({
>         username: email,
>         password: "-",
>       });
> ```
> 
> Error: `"Auth UserPool not configured."` Recovery Suggestion: `"Make sure to call Amplify.configure in your app with userPoolId and userPoolClientId."`
> 
> Clearly I have those there as the configure call without the credentialsProvider works fine. But by adding the credentials provider it fails to load until after the user has signed in. (Which they can't - because sign in fails...)

I'm also seeing a similar issue. Login works until you add a customCredentials provider. When you add a customCredentialsProvider it throws the error: "Auth UserPool not configured".

**Comment by cwomack** [comment]
@jeff2013, can you open a separate issue that details what you're experiencing and any applicable reproduction steps?  I haven't run into any issues using `customCredentialsProvider`, but if you can share some sample code that's causing resulting in this error and how you've set up the Authentication flows/imports then we can try to get you unblocked!  Feel free to create a new issue [here](https://github.com/aws-amplify/amplify-js/issues/new/choose).

**Comment by cwomack** [comment]
@ryanwalters, I'll close this issue out for now since we haven't heard back. If you are still experiencing this, please feel free to reply back and provide any information previously requested (and review [this comment](https://github.com/aws-amplify/amplify-js/issues/13018#issuecomment-2105145393)) and we'd be happy to re-open the issue.

Thank you!

**Comment by ryanwalters** [comment]
@cwomack Apologies for the delayed response.

> @ryanwalters After looking again at the code example in your sample repo, you are configuring the `tokenProvider` and `credentialsProvider` for the server side use cases. However, other parts of your code are only using Amplify on the **_client_** side. Can you clarify if you’re calling any API’s on the server side?

We need Amplify to work on both client and server (this is a Next.js application).

> If you are only intending to use Amplify on the client side, [this document](https://docs.amplify.aws/gen1/nextjs/build-a-backend/auth/manage-user-session/#:~:text=Cookie%20Storage-,CookieStorage,-saves%20the%20tokens) will show you how to configure the `tokenProviders` to change the cookie domain.

Is there a reason `tokenProviders` could not work server side? I see that https://github.com/aws-amplify/amplify-js/issues/11598 has not had much activity since it was opened a year ago.

Ideally, we would have a way to configure a Cognito client that can perform all actions (e.g. sign in, sign out, token refresh, etc) on both the client and the server, including the ability to configure how and where cookies are written.

I would recommend taking a look at how Supabase handles this. They essentially provide functions for both the client and server that allow the end user to specify how to handle getting/setting/removing and configuring cookies.
- https://github.com/supabase/ssr/blob/main/src/createServerClient.ts
- https://github.com/supabase/ssr/blob/main/src/createBrowserClient.ts

Since we were not able to make Amplify 6 work for us, we ended up writing a very basic wrapper around @aws-sdk/client-cognito-identity-provider following a similar API as Supabase's to handle token and cookie management on the server (and middleware) while continuing to use Amplify 5 on the client side, but would obviously rather have an officially supported solution.

**Comment by baalimago** [comment]
Unfortunate that this ticket is closed. It's very much a problem. We have an angular app which we attempted to upgrade from amplify v5 -> v6. Configuration according to the [getting started](https://docs.amplify.aws/gen1/angular/start/getting-started/setup/) documentation which works just fine. But, the project uses federated social logins via identity pool (not user pool), which in turn requires `federatedSignIn`. Since `federatedSignIn` has been deprecated and replaced with [CustomCredentialsProvider](https://docs.amplify.aws/angular/build-a-backend/auth/advanced-workflows/), we're sort of stuck.

In a nutshell, this does not work, throws error `error signing in AuthUserPoolException: Auth UserPool not configured.:
```typescript
Amplify.configure(amplifyconfiguration, {
  Auth: {
    credentialsProvider: customCredentialsProvider
  }
});
```

In fact, this very minimal change breaks `signIn`:
```typescript
Amplify.configure(amplifyconfiguration, {
  Auth: {}
});
```

But changing it to this, makes signin work once more (like it did on v5):
```typescript
Amplify.configure(amplifyconfiguration);
```

So basically ,whenever we try to configure Amplify using any `LibraryOptions`, the user can't sign in. What's odd is that it works prefectly fine to sign up (with verification code) + reset password, only signin doesn't work. 

Since we can't change the federation from identity pool -> user pool without a massive migration headache (a 'migrane', perhaps, if I may be so punny), I fear we'll be stuck using amplify v5 until this has been resolved.

**Comment by lucasantunes-ciandt** [comment]
Same error on our side, but we DO need the custom provider.

Were anyone able to fix it?

**Comment by Samaritan1011001** [comment]
@baalimago Thank you for the detailed comment. May I ask how you are implementing the social login? With federation into identity pool use case, the login largely happens with the social provider (Facebook in your case) then the Social Provider token is given to Amplify so that subsequent calls to AWS services can be authorized. 

Additionally, when you were using the `federatedSignIn` API in V5, were you signing users in via HostedUI? If so, on V6 you can look at the `signInWithRedirect` API.

**Comment by baalimago** [comment]
I'm not longer with the company where I encountered the issue, so I can't doublecheck.

>With federation into identity pool use case, the login largely happens with the social provider (Facebook in your case) then the Social Provider token is given to Amplify so that subsequent calls to AWS services can be authorized.

From what I remember, this is what was the case yes.

> Additionally, when you were using the federatedSignIn API in V5, were you signing users in via HostedUI?

Not a hosted UI in AWS Cognito, no. There was a 'sign in with facebook' button on the "main website" (as in: not in cognito) which started the oauth2 flow.

**Comment by cwomack** [comment]
@baalimago, thank you for the follow up here and closing the loop! 

@lucasantunes-ciandt, we aren't able to reproduce this on our side still.  Do you have any sample code/repo that can reproduce this reliably on the latest (or recent) version of v6?

@ryanwalters, we'll take a look again at the feedback you provided [here](https://github.com/aws-amplify/amplify-js/issues/13018#issuecomment-2261525685) and follow up on this as we investigate further.

**Comment by AllanZhengYP** [comment]
Thanks for providing the reproduction! I can reproduce this issue. Marking it as bug now as I'm investigating the root cause.

**Comment by AllanZhengYP** [comment]
@ryanwalters after deeper investigation into this issue with the sample app provided, I noticed you are using the user pool token provider below:
```js
import { createUserPoolsTokenProvider } from 'aws-amplify/adapter-core';
```

However this `adapter-core` subpath is not suppose to be used directly by users. It's only for library authors who want to build integrations for SSR fameworks. If you are trying to build App with NextJS, you should use the existing package `@aws-amplify/adapter-nextjs`([doc](https://docs.amplify.aws/react/build-a-backend/server-side-rendering/#install-the-amplify-nextjs-adapter))

`createUserPoolsTokenProvider` does not work because it's only suppose to handle token fetch on the server. When you call `signIn` from the client side, the client side conducts additional operations than this token provider is configured for(e.g. fetch device metadata). 

If you are looking for just overwrite the cookie domain on the client-side(browser), instead of touching the auth library options, you should overwrite the default cookie storage for the cognitoTokenProvider[[doc](https://docs.amplify.aws/react/build-a-backend/auth/concepts/tokens-and-credentials/#update-your-token-saving-mechanism)]. 

However, this method does not work for SSR currently. That means even if the cookie is set to higher-level domain, the cookie will NOT be used utilized by the library on server-side. To solve it, we will need to expose the `CookieStorage` options here: https://github.com/aws-amplify/amplify-js/blob/f797dc539f57a55a325b227d8205813b122d7789/packages/aws-amplify/src/initSingleton.ts#L61-L62

@JoeyOverby @jeff2013 The issue you are seeing is different, which caused by overly strict validation. If you still see the issue with the latest version of library, can you open new issues?

**Comment by cwomack** [comment]
Closing this issue as we have not heard back from you. If you are still experiencing this, please feel free to reply back and provide any information previously requested and we'd be happy to re-open the issue.

Thank you!
